### PR TITLE
Waf: refactor

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
@ofborg build wafHook

**Comment by AndersonTorres** [comment]
@ofborg build saldl

**Comment by AndersonTorres** [comment]
@ofborg build saldl

**Comment by AndersonTorres** [comment]
@ofborg build saldl

**Comment by AndersonTorres** [comment]
@ofborg build saldl

File: pkgs/development/tools/build-managers/waf/setup-hook.sh [hunk 1]
```diff
@@ -38,15 +42,11 @@ wafConfigurePhase() {
...
-if [ -z "${dontUseWafConfigure-}" -a -z "${configurePhase-}" ]; then
-    configurePhase=wafConfigurePhase
-fi
-
 wafBuildPhase () {
     runHook preBuild
 
     # set to empty if unset
-    : ${wafFlags=}
+    : "${wafFlags=}"
```
Comment (doronbehar) [inline pkgs/development/tools/build-managers/waf/setup-hook.sh:49]: Are you sure about this?

**Review by doronbehar — COMMENTED** [review]
Looks pretty good to me. Have you successfully tested cross compilation with this?

**Comment by AndersonTorres** [comment]
I tested with some "ofborg building"; as always, the Darwin machines hanged, but the others worked fine.

File: pkgs/development/tools/build-managers/waf/setup-hook.sh (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/waf/setup-hook.sh:49]: Shellcheck complained about the previous form.

Comment (doronbehar) [reply pkgs/development/tools/build-managers/waf/setup-hook.sh:49]: > Shellcheck complained about the previous form.

In general, quoting or not quoting variables makes a difference when these are command arguments. But in this case (just read some Bash docs), it doesn't matter.

**Comment by doronbehar** [comment]
> I tested with some "ofborg building"

What I meant was, building (locally or with ofborg) a package that uses `wafHook`. `fomp` is a good example which you can try, as it has few `buildInputs` that should also be cross compiled. Cross build attributes for `fomp` are for example:

- `pkgsCross.aarch64-unknown-linux-gnu.fomp`
- `pkgsStatic.fomp`
- `pkgsCross.aarch64-unknown-linux-gnu.pkgsStatic.fomp`

See also https://nix.dev/tutorials/cross-compilation .

**Comment by AndersonTorres** [comment]
Let's play, then!

@ofborg build pkgsStatic.fomp

**Comment by doronbehar** [comment]
I think that ofborg's build log indicates that waf's propagated python is not necessarily the python that the host machine is capable of executing. IMO it'll be pretty hard to debug this with ofborg like this, and you'd likely have to do that locally.

**Comment by AndersonTorres** [comment]
waf does not propagate Python.
I have tried to propagate it, but the static build failed.

**Comment by doronbehar** [comment]
Now ofborg's evaluation even fails. Let's put things back to were they were, were things mostly work. Cross compilation should be supported first by waf upstream, and I don't see it mentioned in the [waf book](https://waf.io/book/), so perhaps we should mention that in the documentation as well.

**Comment by doronbehar** [comment]
Another thing to consider is, why do we have both `waf` and `wafHook` attributes. Why not use `waf.setupHook`?

**Comment by AndersonTorres** [comment]
1. using a separate attribute allows opting between the pure `waf` experience and a more batteries-included `wafHook`
2. many projects that use waf include a copy of waf inside their sources. To honor this, the hook verifies if the upstream source already included this, obviating the compilation of waf.
3. about cross compilation, I believe it needs some extra tooling, maybe a module. However I have found only wtools (https://github.com/NixOS/nixpkgs/issues/248127)

File: doc/hooks/waf.section.md [hunk 2]
```diff
@@ -1,3 +1,36 @@
...
+
+In Nixpkgs, `wafHook` overrides the default configure, build, and install phases.
+
+## Variables controlling wafHook
+
+### `wafPath` {#wafPath}
+
+Location of the `waf` tool. It defaults to `./waf`, to honor software projects that include it directly inside their source trees.
+
+If `wafPath` doesn't exist, then `wafHook` will copy the `waf` provided from Nixpkgs to it.
```
Comment (doronbehar) [inline doc/hooks/waf.section.md:13]:
```suggestion
If `wafPath` doesn't exist and `./waf` doesn't exist, then `wafHook` will copy the `waf` provided from Nixpkgs to it.
```

File: doc/hooks/waf.section.md [hunk 3]
```diff
@@ -1,3 +1,36 @@
...
+
+Controls the flags passed to waf tool during configure phase.
+
+### `wafFlags` {#wafFlags}
+
+Controls the flags passed to waf tool during build and install phases.
+
+### `dontAddWafCrossFlags` {#dontAddWafCrossFlags}
+
+When set to `true`, don't add cross compilation flags during configure phase.
```
Comment (doronbehar) [inline doc/hooks/waf.section.md:25]: If we don't have yet cross compilation support, why mention this variable?

File: doc/hooks/waf.section.md [hunk 4]
```diff
@@ -1,3 +1,36 @@
...
+
+### Variables honored by wafHook {#variablesHonoredByWafHook}
+
```
Comment (doronbehar) [inline doc/hooks/waf.section.md:28]: Maybe add here something like:

```markdown
Many variables commonly used by `stdenv.mkDerivation` are also honored by `wafHook`. 
```

File: doc/hooks/waf.section.md [hunk 5]
```diff
@@ -1,3 +1,36 @@
 # wafHook {#wafhook}
 
-Overrides the configure, build, and install phases. This will run the “waf” script used by many projects. If `wafPath` (default `./waf`) doesn’t exist, it will copy the version of waf available in Nixpkgs. `wafFlags` can be used to pass flags to the waf script.
+[Waf](https://waf.io) is a Python-based software building system.
+
+In Nixpkgs, `wafHook` overrides the default configure, build, and install phases.
```
Comment (doronbehar) [inline doc/hooks/waf.section.md:5]: Documenting the _lack_ of cross compilation support is missing, a link to https://github.com/NixOS/nixpkgs/issues/248127 would be nice.

File: doc/hooks/waf.section.md [hunk 6]
```diff
@@ -1,3 +1,36 @@
...
+
+### `wafPath` {#wafPath}
+
+Location of the `waf` tool. It defaults to `./waf`, to honor software projects that include it directly inside their source trees.
+
+If `wafPath` doesn't exist, then `wafHook` will copy the `waf` provided from Nixpkgs to it.
+
+### `wafConfigureFlags` {#wafConfigureFlags}
+
+Controls the flags passed to waf tool during configure phase.
```
Comment (doronbehar) [inline doc/hooks/waf.section.md:17]: Some flags, like `dontUseWafConfigure` are not mentioned.

File: doc/hooks/waf.section.md [hunk 7]
```diff
@@ -1,3 +1,36 @@
...
+### Variables honored by wafHook {#variablesHonoredByWafHook}
+
+- `prefixKey`
+- `configureTargets`
+- `enableParallelBuilding`
+- `enableParallelInstalling`
+- `buildFlags`
+- `buildTargets`
+- `installFlags`
+- `installTargets`
```
Comment (doronbehar) [inline doc/hooks/waf.section.md:36]: Waf uses `make` under the hood? Perhaps that's worth mentioning as well.

File: pkgs/development/tools/build-managers/waf/default.nix [hunk 8]
```diff
@@ -1,44 +1,71 @@
...
   };
 
-  nativeBuildInputs = [ python3 ensureNewerSourcesForZipFilesHook ];
+  nativeBuildInputs = [
+    ensureNewerSourcesForZipFilesHook
+    python3
+  ];
+
+  buildInputs = [
+    # waf executable uses `#!/usr/bin/env python`
```
Comment (doronbehar) [inline pkgs/development/tools/build-managers/waf/default.nix:27]: So this makes the executables there `patchShebangs` automatically? Please continue the sentence in the comment.

File: pkgs/development/tools/build-managers/waf/default.nix [hunk 9]
```diff
@@ -1,44 +1,71 @@
...
+    python3
+  ];
 
-  # waf bin has #!/usr/bin/env python
-  buildInputs = [ python3 ];
+  strictDeps = true;
```
Comment (doronbehar) [inline pkgs/development/tools/build-managers/waf/default.nix:31]: Isn't this the default?

**Review by doronbehar — DISMISSED** [review]

File: doc/hooks/waf.section.md (hunk 2 above)
Comment (AndersonTorres) [reply doc/hooks/waf.section.md:13]: According to the hook script (and reflected by this documentation), wafPath is set to `./waf` when not set by the user. So, that suggestion would be redundant.

Further, the hook script does not try to find a "hardcoded" path, it only verifies if `wafPath` exists.

It is useful when, by any reason, the upstream-provided waf is not desired: just set `wafPath = ./nixpkgs-waf` and be happy.

File: pkgs/development/tools/build-managers/waf/default.nix (hunk 9 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/waf/default.nix:31]: This is just a line moving around.

Also, as far as I remember, strictDeps is not default on Nixpkgs. There is even a warning on NixOS about mass rebuilds when it is set globally.

Comment (doronbehar) [reply pkgs/development/tools/build-managers/waf/default.nix:31]: > This is just a line moving around.
> 
> Also, as far as I remember, strictDeps is not default on Nixpkgs. There is even a warning on NixOS about mass rebuilds when it is set globally.

Fair enough.

File: doc/hooks/waf.section.md (hunk 5 above)
Comment (AndersonTorres) [reply doc/hooks/waf.section.md:5]: Curiously, DDGing a bit I found that Samba does explain a bit about cross-compilation:

https://wiki.samba.org/index.php/Waf#cross-compiling

https://wiki.samba.org/index.php/BuildsystemUseAndWhy

File: doc/hooks/waf.section.md (hunk 7 above)
Comment (doronbehar) [reply doc/hooks/waf.section.md:36]: I read now the rest of `waf/setup-hook.sh`, and I don't see a reason why would we support for instance both `installFlags` and `wafFlags`. How about supporting only `wafInstallFlags` and `wafBuildFlags`?

Comment (AndersonTorres) [reply doc/hooks/waf.section.md:36]: You say, all flags should be  `waf`-prefixed?  Why?

The `wafFlags` are common  flags to be passed to build and install.

On the other hand, the script is a bit inconsistent.

File: pkgs/development/tools/build-managers/waf/default.nix (hunk 8 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/waf/default.nix:27]: No, it just makes Python available at build time.

**Comment by doronbehar** [comment]
@AndersonTorres I don't have enough time to continue discuss the topics presented here, and I also lack motivation as I barely interact with waf on Nixpkgs, so I'm unsubscribing from this PR. Good luck with that :)

**Comment by AndersonTorres** [comment]
No problem. Thanks for the help!

**Comment by AndersonTorres** [comment]
This is  fine for a refactor.  I will work on the cross questions later.

**Comment by AndersonTorres** [comment]
Since 2.0.26 was released, I will merge this and continue later.

**Pull request merged**